### PR TITLE
permissions: ensure returned conditions match permission in `PermissionPolicy#handle`

### PR DESCRIPTION
**Pull request by mtlewis**
## Hey, I just made a Pull Request!

Adds some extra typing to the helpers for creating conditional decisions, which provide some type checks to ensure that conditions for resources of type `x` can only be returned when handling `ResourcePermissions` for that same resource type.

### Old Syntax

```typescript 
  handle(
    request: PolicyAuthorizeQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    // Handle a specific permission by name
    if (request.permission.name === 'catalog.entity.delete') {
      // Hope I got the name right!
      return createCatalogPolicyDecision(
        catalogConditions.hasSpec('type', 'website'),
      );
    }

    // Handle all catalog-entity permissions
    if (request.permission.resourceType === RESOURCE_TYPE_CATALOG_ENTITY) {
      // Hope I got the resourceType right!
      return createCatalogPolicyDecision(
        catalogConditions.hasSpec('type', 'website'),
      );
    }

    /* ... */
  }
```

### New Syntax

```typescript
  async handle(
    request: PolicyQuery,
    _user?: BackstageIdentityResponse,
  ): Promise<PolicyDecision> {
    // Handle a specific permission by name
    if (isPermission(request.permission, catalogEntityDeletePermission)) {
      // Type gets narrowed to ResourcePermission<'catalog-entity'>.
      // Both the permission and conditions supplied to 
      // `createCatalogConditionalDecision` must have the 'catalog-entity'
      // resource type.
      return createCatalogConditionalDecision(
        request.permission,
        catalogConditions.hasSpec('type', 'website'),
      );
    }

    // Handle all catalog-entity permissions
    if (
      isResourcePermission(request.permission, RESOURCE_TYPE_CATALOG_ENTITY)
    ) {
      // Type gets narrowed to ResourcePermission<'catalog-entity'>
      return createCatalogConditionalDecision(
        request.permission,
        catalogConditions.hasSpec('type', 'website'),
      );
    }

    /* ... */
  }
```

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Joonpark13 — APPROVED** [review]
Looks great! I think this will be a very helpful change to the API.

File: .changeset/beige-avocados-matter.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+'@backstage/plugin-catalog-backend': patch
+---
+
+Replace `createCatalogPolicyDecision` export with `createCatalogConditionalDecision`, which accepts a permission parameter of type `ResourcePermission<'catalog-entity'>` along with conditions. The permission passed is expected to be the handled permission in `PermissionPolicy#handle`, whose type must first be narrowed using methods like `isPermission` and `isResourcePermission`:
```
Comment (Rugvip) [inline .changeset/beige-avocados-matter.md:5]: Could you highlight that this is a breaking change of an alpha API? Both to signal the breakage but also to explain why it's done without a major bump

**Review by Rugvip — APPROVED** [review]
Looks solid 👌

**Comment by mtlewis** [comment]
@joeporpeglia is kindly going to step in to wrap up the last bits of feedback on this PR, since I'll be away for a while.

File: .changeset/beige-avocados-matter.md (hunk 1 above)
Comment (joeporpeglia) [reply .changeset/beige-avocados-matter.md:5]: Do we have a specific way to note breaking alpha apis? Perhaps `**BREAKING (alpha api):**`

**Comment by backstage-service** [comment]
## Changed Packages

| Package Name | Package Path | Changeset Bump | Current Version |
|:-------------|:-------------|:--------------:|:----------------|
| @backstage/plugin-catalog-backend | plugins/catalog-backend | **minor** | `v1.0.1-next.0` |
| @backstage/plugin-permission-common | plugins/permission-common | **minor** | `v0.5.3` |
| @backstage/plugin-permission-node | plugins/permission-node | **minor** | `v0.5.6-next.0` |

**Comment by backstage-service** [comment]
## Changed Packages

| Package Name | Package Path | Changeset Bump | Current Version |
|:-------------|:-------------|:--------------:|:----------------|
| @backstage/plugin-catalog-backend | plugins/catalog-backend | **minor** | `v1.0.1-next.0` |
| @backstage/plugin-permission-common | plugins/permission-common | **minor** | `v0.5.3` |
| @backstage/plugin-permission-node | plugins/permission-node | **minor** | `v0.5.6-next.0` |

**Pull request merged**